tntnet:  error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'

Building `tntnet-2.2.1` ([website](http://www.tntnet.org/)) on Linux with latest `zlib-ng` (5bd3c582908503135fd8d6326ca0e2a6a947b44f from develop branch) and `--zlib-compat` for x86_64 host (not cross-compiling) results in the following build failure:
```
libtool: compile:  /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/toolchain/bin/host-g++ -DHAVE_CONFIG_H -I. -I/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common -Wno-long-long -Wall -pedantic -DPKGLIBDIR=\"/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/toolchain/lib/tntnet\" -march=native -O2 -Wall -pipe -I/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/toolchain/include -Wno-format-security -MT libtntnet_la-httpreply.lo -MD -MP -MF .deps/libtntnet_la-httpreply.Tpo -c /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/httpreply.cpp  -fPIC -DPIC -o .libs/libtntnet_la-httpreply.o
libtool: compile:  /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/toolchain/bin/host-gcc -DHAVE_CONFIG_H -I. -I/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common -Wno-long-long -Wall -pedantic -march=native -O2 -Wall -pipe -I/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/toolchain/include -Wno-format-security -MT ioapi.lo -MD -MP -MF .deps/ioapi.Tpo -c /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c  -fPIC -DPIC -o .libs/ioapi.o
In file included from /home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:17:0:
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:127:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef voidpf   (ZCALLBACK *open_file_func)      OF((voidpf opaque, const char* filename, int mode));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:128:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef uLong    (ZCALLBACK *read_file_func)      OF((voidpf opaque, voidpf stream, void* buf, uLong size));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:129:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef uLong    (ZCALLBACK *write_file_func)     OF((voidpf opaque, voidpf stream, const void* buf, uLong size));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:130:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef int      (ZCALLBACK *close_file_func)     OF((voidpf opaque, voidpf stream));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:131:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef int      (ZCALLBACK *testerror_file_func) OF((voidpf opaque, voidpf stream));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:133:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef long     (ZCALLBACK *tell_file_func)      OF((voidpf opaque, voidpf stream));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:134:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef long     (ZCALLBACK *seek_file_func)      OF((voidpf opaque, voidpf stream, uLong offset, int origin));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:140:5: error: unknown type name 'open_file_func'
     open_file_func      zopen_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:141:5: error: unknown type name 'read_file_func'
     read_file_func      zread_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:142:5: error: unknown type name 'write_file_func'
     write_file_func     zwrite_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:143:5: error: unknown type name 'tell_file_func'
     tell_file_func      ztell_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:144:5: error: unknown type name 'seek_file_func'
     seek_file_func      zseek_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:145:5: error: unknown type name 'close_file_func'
     close_file_func     zclose_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:146:5: error: unknown type name 'testerror_file_func'
     testerror_file_func zerror_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:150:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef ZPOS64_T (ZCALLBACK *tell64_file_func)    OF((voidpf opaque, voidpf stream));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:151:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef long     (ZCALLBACK *seek64_file_func)    OF((voidpf opaque, voidpf stream, ZPOS64_T offset, int origin));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:152:51: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 typedef voidpf   (ZCALLBACK *open64_file_func)    OF((voidpf opaque, const void* filename, int mode));
                                                   ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:156:5: error: unknown type name 'open64_file_func'
     open64_file_func    zopen64_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:157:5: error: unknown type name 'read_file_func'
     read_file_func      zread_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:158:5: error: unknown type name 'write_file_func'
     write_file_func     zwrite_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:159:5: error: unknown type name 'tell64_file_func'
     tell64_file_func    ztell64_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:160:5: error: unknown type name 'seek64_file_func'
     seek64_file_func    zseek64_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:161:5: error: unknown type name 'close_file_func'
     close_file_func     zclose_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:162:5: error: unknown type name 'testerror_file_func'
     testerror_file_func zerror_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:166:28: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 void fill_fopen64_filefunc OF((zlib_filefunc64_def* pzlib_filefunc_def));
                            ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:167:26: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 void fill_fopen_filefunc OF((zlib_filefunc_def* pzlib_filefunc_def));
                          ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:173:5: error: unknown type name 'open_file_func'
     open_file_func      zopen32_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:174:5: error: unknown type name 'tell_file_func'
     tell_file_func      ztell32_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:175:5: error: unknown type name 'seek_file_func'
     seek_file_func      zseek32_file;
     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:186:21: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 voidpf call_zopen64 OF((const zlib_filefunc64_32_def* pfilefunc,const void*filename,int mode));
                     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:187:22: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 long    call_zseek64 OF((const zlib_filefunc64_32_def* pfilefunc,voidpf filestream, ZPOS64_T offset, int origin));
                      ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.h:188:23: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 ZPOS64_T call_ztell64 OF((const zlib_filefunc64_32_def* pfilefunc,voidpf filestream));
                       ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c: In function 'call_zopen64':
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:21:46: warning: comparison between pointer and integer
     if (pfilefunc->zfile_func64.zopen64_file != NULL)
                                              ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:22:17: error: invalid type argument of unary '*' (have 'int')
         return (*(pfilefunc->zfile_func64.zopen64_file)) (pfilefunc->zfile_func64.opaque,filename,mode);
                 ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:25:17: error: invalid type argument of unary '*' (have 'int')
         return (*(pfilefunc->zopen32_file))(pfilefunc->zfile_func64.opaque,(const char*)filename,mode);
                 ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c: In function 'call_zseek64':
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:31:46: warning: comparison between pointer and integer
     if (pfilefunc->zfile_func64.zseek64_file != NULL)
                                              ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:32:17: error: invalid type argument of unary '*' (have 'int')
         return (*(pfilefunc->zfile_func64.zseek64_file)) (pfilefunc->zfile_func64.opaque,filestream,offset,origin);
                 ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:39:21: error: invalid type argument of unary '*' (have 'int')
             return (*(pfilefunc->zseek32_file))(pfilefunc->zfile_func64.opaque,filestream,offsetTruncated,origin);
                     ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c: In function 'call_ztell64':
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:45:46: warning: comparison between pointer and integer
     if (pfilefunc->zfile_func64.zseek64_file != NULL)
                                              ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:46:17: error: invalid type argument of unary '*' (have 'int')
         return (*(pfilefunc->zfile_func64.ztell64_file)) (pfilefunc->zfile_func64.opaque,filestream);
                 ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:49:29: error: invalid type argument of unary '*' (have 'int')
         uLong tell_uLong = (*(pfilefunc->ztell32_file))(pfilefunc->zfile_func64.opaque,filestream);
                             ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c: In function 'fill_zlib_filefunc64_32_def_from_filefunc32':
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:59:48: warning: assignment makes integer from pointer without a cast [-Wint-conversion]
     p_filefunc64_32->zfile_func64.zopen64_file = NULL;
                                                ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:64:48: warning: assignment makes integer from pointer without a cast [-Wint-conversion]
     p_filefunc64_32->zfile_func64.ztell64_file = NULL;
                                                ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:65:48: warning: assignment makes integer from pointer without a cast [-Wint-conversion]
     p_filefunc64_32->zfile_func64.zseek64_file = NULL;
                                                ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c: At top level:
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:75:42: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 static voidpf  ZCALLBACK fopen_file_func OF((voidpf opaque, const char* filename, int mode));
                                          ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:76:42: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 static uLong   ZCALLBACK fread_file_func OF((voidpf opaque, voidpf stream, void* buf, uLong size));
                                          ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:77:43: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 static uLong   ZCALLBACK fwrite_file_func OF((voidpf opaque, voidpf stream, const void* buf,uLong size));
                                           ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:78:45: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 static ZPOS64_T ZCALLBACK ftell64_file_func OF((voidpf opaque, voidpf stream));
                                             ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:79:44: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 static long    ZCALLBACK fseek64_file_func OF((voidpf opaque, voidpf stream, ZPOS64_T offset, int origin));
                                            ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:80:43: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 static int     ZCALLBACK fclose_file_func OF((voidpf opaque, voidpf stream));
                                           ^
/home/ubuntu/projects/LibreELEC.tv/build.LibreELEC-Generic.x86_64-9.80-devel/tntnet-2.2.1/framework/common/ioapi.c:81:43: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'OF'
 static int     ZCALLBACK ferror_file_func OF((voidpf opaque, voidpf stream));
                                           ^
```

`tntnet` build log: http://ix.io/22Qu

No such build issue with upstream `zlib-1.2.11`.

Using `gcc-9.2.0`.

I don't currently have a fix for this, but happy to test any proposals.

`OF()` macro is also needed when cross-compiling `libretro-uae4arm` ([website](https://github.com/libretro/uae4arm-libretro)) for arm:
```
In file included from src/zfile_archive.cpp:14:
src/archivers/zip/unzip.h:123:45: error: expected initializer before 'OF'
  123 | extern int ZEXPORT unzStringFileNameCompare OF ((const char* fileName1,
      |                                             ^~
src/archivers/zip/unzip.h:136:32: error: expected initializer before 'OF'
  136 | extern unzFile ZEXPORT unzOpen OF((struct zfile *path));
      |                                ^~
src/archivers/zip/unzip.h:147:29: error: expected initializer before 'OF'
  147 | extern int ZEXPORT unzClose OF((unzFile file));
      |                             ^~
src/archivers/zip/unzip.h:154:37: error: expected initializer before 'OF'
  154 | extern int ZEXPORT unzGetGlobalInfo OF((unzFile file,
      |                                     ^~
...
src/zfile_archive.cpp: In function 'void archive_close_zip(void*)':
src/zfile_archive.cpp:203:5: error: 'unzClose' was not declared in this scope
  203 |     unzClose(handle);
      |     ^~~~~~~~
src/zfile_archive.cpp: In function 'zvolume* archive_directory_zip(zfile*)':
src/zfile_archive.cpp:214:10: error: 'unzOpen' was not declared in this scope
  214 |     uz = unzOpen (z);
      |          ^~~~~~~
```

build log: http://ix.io/22QR

How difficult would it be to restore this macro and thus improve compatibility with `zlib`? It's not a modern compiler issue, as this is building with `gcc-9.2.0`.

--- a/zconf.h.in
+++ b/zconf.h.in
@@ -55,6 +55,10 @@
                         /* Type declarations */
 
 
+#ifndef OF /* function prototypes */
+#  define OF(args)  args
+#endif
+
 #if defined(WINDOWS) || defined(WIN32)
    /* If building or using zlib as a DLL, define ZLIB_DLL.
     * This is not mandatory, but it offers a little performance increase.

